packages/google_sign_in/example/lib/main.dart

```diff
@@ -94,18 +97,45 @@ class SignInDemoState extends State<SignInDemo> {
     return null;
   }
 
-  Future<Null> _handleSignIn() async {
-    try {
-      await _googleSignIn.signIn();
-    } catch (error) {
-      print(error);
-    }
+  void _handleSignIn() {
+    _handleGoogleSignIn()
+        .then((GoogleSignInAccount user) => _afterSignIn)
+        .catchError((dynamic error) => _onSignInError);
```

Comment:
Certainly an improvement but again much less readable because the flow goes backwards w.r.t. how the code is executed. await makes async code look more like normal control flow. 

The code you have quotes certainly achieves the same thing but looks very far from normal control flow. For instance, I see `(e) =>` first, then I have to keep reading to see what type of `e` this is. I see `_onSignIn` first but that's actually not the next thing that executes after `handleSignIn()`  (since it is async).

I admit for a certain group of developers, `.then` and `.catchError` makes a lot of sense because they are used to the pattern. But we found that it is an expensive habit to form, especially when people start chaining:

```dart
myFuture
    .then((a) => new B(a))
    .then((b) => new C(b))
    .then((c) => print(c))
    .catchError((e) => print(e));
```

We had an actual bug in our code because someone did this:
```dart

myFuture
    .then((a) => new B(a))
    .catchError((e) => print(e))
    .then((b) => new C(b))
    .then((c) => print(c));
```
which works fine until you get an error somewhere down the chain.